rtrlib/aspa/aspa_array/aspa_array.c

```diff
@@ -0,0 +1,256 @@
+				if (array->data[i].provider_asns) {
+					lrtr_free(array->data[i].provider_asns);
+					array->data[i].provider_asns = NULL;
+				}
+			}
+		}
+
+		lrtr_free(array->data);
+	}
+
+	lrtr_free(array);
+}
+
+// MARK: - Manipulation
+
+static enum aspa_status aspa_array_reallocate(struct aspa_array *array)
+{
+	// the factor by how much the capacity will increase: new_capacity = old_capacity * SIZE_INCREASE_EXPONENTIAL
+	const size_t SIZE_INCREASE_EXPONENTIAL = 2;
+
+	// allocation the new chunk of memory
+	struct aspa_record *tmp =
+		lrtr_realloc(array->data, sizeof(struct aspa_record) * array->capacity * SIZE_INCREASE_EXPONENTIAL);
+
+	// malloc failed so returning an error
+	if (!tmp)
+		return ASPA_ERROR;
+
+	array->data = tmp;
+	array->capacity *= SIZE_INCREASE_EXPONENTIAL;
+	return ASPA_SUCCESS;
+}
+
+static enum aspa_status aspa_array_insert_internal(struct aspa_array *array, size_t index, struct aspa_record *record,
+						   bool copy_providers)
+{
+	// check if this element will fit into the array
+	if (array->size >= array->capacity) {
+		// increasing the array's size so the new element fits
+		if (aspa_array_reallocate(array) != ASPA_SUCCESS)
+			return ASPA_ERROR;
+	}
+
+	uint32_t *provider_asns = NULL;
+
+	if (record->provider_count > 0) {
+		if (copy_providers) {
+			size_t provider_size = record->provider_count * sizeof(uint32_t);
+
+			provider_asns = lrtr_malloc(provider_size);
+			if (!provider_asns)
+				return ASPA_ERROR;
+
+			memcpy(provider_asns, record->provider_asns, provider_size);
+		} else {
+			provider_asns = record->provider_asns;
+		}
+	}
+
+	// No need to move if last element
+	if (index < array->size) {
+		size_t trailing = (array->size - index) * sizeof(struct aspa_record);
+
+		/*               trailing
+		 *		     /-------------\
+		 * #3 #8 #11 #24 #30 #36 #37
+		 * #3 #8 #11  *  #24 #30 #36 #37
+		 *            ^   ^
+		 *		index   index + 1
+		 */
+		memmove(&array->data[index + 1], &array->data[index], trailing);
+	}
+
+	// append the record at the end
+	array->data[index] = *record;
+	array->data[index].provider_asns = provider_asns;
+	array->size += 1;
+
+	return ASPA_SUCCESS;
+}
+
+enum aspa_status aspa_array_insert(struct aspa_array *array, size_t index, struct aspa_record *record,
+				   bool copy_providers)
+{
+	if (!array || index > array->size || !array->data || !record ||
+	    ((record->provider_count > 0) && !record->provider_asns))
+		return ASPA_ERROR;
+
+	return aspa_array_insert_internal(array, index, record, copy_providers);
+}
+
+enum aspa_status aspa_array_append(struct aspa_array *array, struct aspa_record *record, bool copy_providers)
+{
+	if (!array || !array->data || !record || ((record->provider_count > 0) && !record->provider_asns))
+		return ASPA_ERROR;
+
+	return aspa_array_insert_internal(array, array->size, record, copy_providers);
+}
+
+enum aspa_status aspa_array_remove(struct aspa_array *array, size_t index, bool free_providers)
```

Comment:
If I understand it correctly the allocated memory for the array is currently not reduced when only remove operations are done.
Why is this not done and needed?
Can't this lead to a situation where the array uses much memory then actually needed for a long period of time?

Reply:
This could be, but I guess is unlikely. 

Do you think it is necessary to scale it down again?

Reply:
To me it is still unclear why a dynamic resizing array is a better choice then a fixed size array, that would only allocate as much space as needed.

> Do you think it is necessary to scale it down again?

As I understand, in the current implementation we always allocate `2 * old_capacity` elements when more space is needed.
Therefore we can end up with a lot of unused space. For example, when the array holds 2048 elements and we add/remove 1 additional one we end up with having space for 4096 elements allocated unnecessarily for a long time.
In the worst scenario this could even lead to that aspa_array holds more unused memory then rtrlib uses effectively. I think we should make it more efficient and prevent that aspa_array can end up holding so much unused memory.

Reply:
@fho I implemented also now a scale down now in the `aspa_array_remove` function, so the capacity is never larger then twice the size of the array.

Reply:
@fho Out of interest, wouldn't have lowering the factor also solved this? I was worried about the number of allocations, but again, this depends on the number and size of updates we get..

Reply:
Its a trade of. The lower the factor, the more often we need to reallocate the higher the factor, the more wasted memory we potentially have.

Reply:
> @fho Out of interest, wouldn't have lowering the factor also solved this? I was worried about the number of allocations, but again, this depends on the number and size of updates we get..

Also with a lower factor we might end up in the situation that we add a lot of records to the array, then remove the majority again and never need the memory again:

Exaggeration:
- we add 10.000 aspa records,
- we remove 9.000 aspa records, never need more space then for 1000 elements again,
=> we have allocated unnecessarily space for 9.000 elements until the program terminates


The factor of 2 also seems to big to me, the bigger the array grows the more unlikely it becomes that we ever use the reserved space.